### PR TITLE
ENYO-5480: Prevent unexpected scrolling when updated by app

### PR DESCRIPTION
**Pull request by ybsung**
### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)

`onUpdate` prop in moonstone/Scroller.componentDidUpdate() is designed to be called after ui/Scrollable.enqueueForceUpdate() is called. But `this.setState()` from app can also execute this function and lead Scroller to scroll unexpectedly.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

I prevent calling `onUpdate` in `ui/Scrollable.componentDidUpdate()` by adding the `shouldUpdateScroller` flag. The `shouldUpdateScroller` flag could be set to `true` when `ui/Scrollable.enqueueForceUpdate()` is called.

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

If `onUpdate` is called from Scroller, the `onUpdate` could be called twice.
The first time when calling `this.forceUpdate()` in `ui/Scrollable.enqueueForceUpdate()`.
The second time when calling `this.setState()` in `ui/Scrollable.componentDidUpdate()` to update scrollbar visibility.

So I moved `onUpdate()` from moonstone/Scroller to ui/Scrollable and called it depending on the visibility change of the scrollbar

In addition, I removed the Scroller test case related with `onUpdate`.

### Links
[//]: # (Related issues, references)

ENYO-5480

### Comments

Enact-DCO-1.0-Signed-off-by: YB Sung (yb.sung@lge.com)

### CODE REVIEW

File: packages/moonstone/Scrollable/Scrollable.js [hunk 1]
```diff
@@ -548,8 +548,9 @@ class ScrollableBase extends Component {
 
 	// Callback for scroller updates; calculate and, if needed, scroll to new position based on focused item.
 	handleScrollerUpdate = () => {
-		if (this.uiRef.scrollToInfo === null && Spotlight.getPointerMode()) {
+		if (this.uiRef.scrollOnFocusInPointerMode && this.uiRef.scrollToInfo === null && Spotlight.getPointerMode()) {
```
Comment (MikyungKim) [inline packages/moonstone/Scrollable/Scrollable.js]: Can't we move these conditions to ui/Scrollable?

File: packages/moonstone/Scrollable/Scrollable.js [hunk 2]
```diff
@@ -548,8 +548,9 @@ class ScrollableBase extends Component {
...
 			this.calculateAndScrollTo();
+			this.uiRef.scrollOnFocusInPointerMode = false;
```
Comment (MikyungKim) [inline packages/moonstone/Scrollable/Scrollable.js]: This one also.

File: packages/ui/Scrollable/Scrollable.js [hunk 3]
```diff
@@ -259,6 +259,14 @@ class ScrollableBase extends Component {
 		 */
 		onScrollStop: PropTypes.func,
 
+		/**
+		 * Called when invalidating [Scroller]{@link moonstone/Scroller.Scroller}'s bounds
```
Comment (MikyungKim) [inline packages/ui/Scrollable/Scrollable.js]: Shouldn't it be ui/Scroller?

File: packages/ui/Scrollable/Scrollable.js [hunk 4]
```diff
@@ -406,8 +415,8 @@ class ScrollableBase extends Component {
 		) {
 			this.deferScrollTo = false;
 			this.isUpdatedScrollThumb = this.updateScrollThumbSize();
-		} else {
-			this.updateScrollbars();
+		} else if (!this.updateScrollbars() && onUpdate && !this.childRef.hasOwnProperty('hasDataSizeChanged')) {
```
Comment (MikyungKim) [inline packages/ui/Scrollable/Scrollable.js]: Let's refactor this part based on more trimmed logic.

File: packages/ui/Scrollable/Scrollable.js [hunk 5]
```diff
@@ -446,6 +455,7 @@ class ScrollableBase extends Component {
 	// TODO: consider replacing forceUpdate() by storing bounds in state rather than a non-
 	// state member.
 	enqueueForceUpdate = () => {
+		this.scrollOnFocusInPointerMode = true;
```
Comment (MikyungKim) [inline packages/ui/Scrollable/Scrollable.js]: Let's rename it to... `shouldUpdateScroller`?

File: packages/ui/Scrollable/ScrollableNative.js [hunk 6]
```diff
@@ -252,6 +252,14 @@ class ScrollableBaseNative extends Component {
 		 */
 		onScrollStop: PropTypes.func,
 
+		/**
+		 * Called when invalidating [Scroller]{@link ui/Scroller.Scroller}'s bounds
```
Comment (MikyungKim) [inline packages/ui/Scrollable/ScrollableNative.js]: ScrollerNative?

File: packages/moonstone/CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,7 @@ The following is a curated list of changes in the Enact moonstone module, newest
 
 ### Fixed
 
+- `moonstone/Scroller` to prevent scrolling when hovering the truncated spottable item
```
Comment (MikyungKim) [inline packages/moonstone/CHANGELOG.md]: please update the log

File: packages/ui/Scrollable/Scrollable.js [hunk 8]
```diff
@@ -404,10 +414,15 @@ class ScrollableBase extends Component {
 			hasDataSizeChanged === false &&
 			(isHorizontalScrollbarVisible && !prevState.isHorizontalScrollbarVisible || isVerticalScrollbarVisible && !prevState.isVerticalScrollbarVisible)
 		) {
+			// For VirtualList
 			this.deferScrollTo = false;
 			this.isUpdatedScrollThumb = this.updateScrollThumbSize();
-		} else {
-			this.updateScrollbars();
+		} else if (!this.updateScrollbars()) {
+			if (this.shouldUpdateScroller === true && !this.childRef.hasOwnProperty('hasDataSizeChanged')) {
```
Comment (MikyungKim) [inline packages/ui/Scrollable/Scrollable.js]: let's remove `=== true`

**Comment by MikyungKim** [comment]
Looks okay. I think we need to revisit this later after refactoring `enqueueForceUpdate` and logic in `componentDidUpdate`.

**Comment by teckliew** [comment]
:lgtm:

**Comment by ryanjduffy** [comment]
I'd like to better understand the necessity of this change before moving ahead. I've done some analysis (which I've included below) but recognize that the others on this PR have a deeper knowledge of this codebase than I

### Analysis
I noticed a disconnect between the original requirement for adding `onUpdate` and the way in which it was implemented. The requirement surfaced due to a need to scroll a Component (in that case, the entire ExpandableList) into view when that component caused a change in scroll bounds. It appears in the original implementation that we asserted that nearly any update of the component should cause this behavior but discovered that updates introduced upstream (as was described in this ticket) would inadvertently cause scrolling.

The gap in my view is that we didn't verify that the scroll bounds changed before invoking this logic. I adjusted the callback with the following and it seems to resolve the issue but I'm not sure it's a complete or appropriate solution.

```
	// Callback for scroller updates; calculate and, if needed, scroll to new position based on focused item.
	handleScrollerUpdate = () => {
		if (this.uiRef.scrollToInfo === null && Spotlight.getPointerMode()) {
			const scrollHeight = this.uiRef.getScrollBounds().scrollHeight;
			if (scrollHeight !== this.uiRef.bounds.scrollHeight) {
				this.calculateAndScrollTo();
			}
		}
	}
```

**Comment by ryanjduffy** [comment]
Closing in favor of #1873

**Pull request closed without merging**